075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+* before scaling the cluster down, in order to rebalance the cluster and move partition replicas off the brokers to be removed, the user has to manually create a `KafkaRebalance` custom resource using the `spec.mode: remove-brokers` mode with the IDs of the brokers to be removed set on the `spec.brokers` field.
+
+For more details, the proposal [035](https://github.com/strimzi/proposals/blob/main/035-rebalance-types-scaling-brokers.md) introduced the rebalancing modes mentioned above, which will be also referenced across this proposal.
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two steps process.
+
+* scaling the cluster up by adding brokers, then running a rebalance.
+* running a rebalance, then scaling the cluster down by removing brokers.
+
+In order to simplify the process, it would be great for the user to deal with just scaling the cluster and then having the operator taking care of running a rebalancing operation automatically when needed.
+
+## Proposal
+
+This proposal is about providing the user the possibility to automatically run a rebalancing when they scale the cluster up (by adding brokers) or scale the cluster down (by removing brokers) by changing the `spec.replicas` within a `Kafka` or `KafkaNodePool` custom resource.
+
+### Extending the `Kafka.spec.cruiseControl` field with auto-rebalancing related configuration
+
+The `Kafka` custom resource is extended by adding a new `spec.cruiseControl.autoRebalance` field.
+By default, without such field, no auto-rebalancing runs on scaling.
+
+The `spec.cruiseControl.autoRebalance` field allows to specify a list of rebalancing modes with the corresponding configuration to be used.
+For each entry in the list, the mode is defined by the `spec.cruiseControl.autoRebalance.mode` field (i.e. `add-brokers` or `remove-brokers`) and the corresponding rebalancing configuration is defined as a reference to a `KafkaRebalanceSettings` custom resource by using the `spec.cruiseControl.autoRebalance.settings` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+
+In the following example, a user can decide to use the same rebalancing settings when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the same rebalancing settings when adding or removing brokers
+      - mode: add-brokers
+        settings: my-add-remove-brokers-rebalancing-settings
+      - mode: remove-brokers
+        settings: my-add-remove-brokers-rebalancing-settings
```

Comment:
Have you considered how to make it enabled by default? At least for scale-down, it seems like there are not many reasons why should it be disabled by default (at least in long term).

Reply:
But an auto-rebalancing enabled by default would need Cruise Control enabled by default which is something we don't have. Not sure how much of a good UX it would be saying "if you enable Cruise Control, you get autorebalancing by default on scaling". Not that bad, not that good imho. Furthermore, rebalancing could take long so if enabled by default on scaling up, it could mean longer time to scale up (even if the user doesn't need spreading current partitions on new brokers). If enabled by default on scaling down, it could mean longer time as well but not warning the user that the brokers they are removing are hosting partitions which will be moved. I think having control and warnings in this cases is better. It's always simpler adding a configuration to enable something (which is disabled by default) than disabling the ... disabled by default (see our TLS-enabled by default intra cluster communication for example).

Reply:
I do not follow why would it take longer. The scale-down does not happen when the nodes are empty. So it would actually "make it happen". For scale-up, it happens after the scale-up, so the user might not care for it. Or at least my expectation was that the rebalancing will be asynchronous and not blocking the reconciliation, maybe I got it wrong.

Reply:
> The scale-down does not happen when the nodes are empty. So it would actually "make it happen".

Yes. But if nodes are not empty, the user could argue "why scaling down is not happening" ... the reason is because we are running a rebalancing by default to take partitions off those nodes. I think it's good if the user explicitly activated the auto-rebalancing (not having it by default).

> For scale-up, it happens after the scale-up, so the user might not care for it

That's true but user should be also aware that clients could get errors because of leaders' partitions moving around due to the rebalancing. So again, a rebalancing which the user doesn't expect because he didn't explicitly said to do so.

> Or at least my expectation was that the rebalancing will be asynchronous and not blocking the reconciliation, maybe I got it wrong.

You are not wrong. My latest PR https://github.com/strimzi/strimzi-kafka-operator/pull/9940 did exactly this. Now the rebalancing happens across reconciliation not blocking the current one (as it was before).

Reply:
> > The scale-down does not happen when the nodes are empty. So it would actually "make it happen".
> 
> Yes. But if nodes are not empty, the user could argue "why scaling down is not happening" ... the reason is because we are running a rebalancing by default to take partitions off those nodes. I think it's good if the user explicitly activated the auto-rebalancing (not having it by default).
> 
> > For scale-up, it happens after the scale-up, so the user might not care for it
> 
> That's true but user should be also aware that clients could get errors because of leaders' partitions moving around due to the rebalancing. So again, a rebalancing which the user doesn't expect because he didn't explicitly said to do so.

Going back to the original idea of the comment -> do you really think this should be opt-in forever? I can understand your concerns. But at the same time, I would think that this would be something I would want to have eventualy turned on by default.

> > Or at least my expectation was that the rebalancing will be asynchronous and not blocking the reconciliation, maybe I got it wrong.
> 
> You are not wrong. My latest PR [strimzi/strimzi-kafka-operator#9940](https://github.com/strimzi/strimzi-kafka-operator/pull/9940) did exactly this. Now the rebalancing happens across reconciliation not blocking the current one (as it was before).

But I'm talking about the Kafka resource reconciliation here. Do you expect it to happen synchrinously as part of the Kafka reconciliation? I do not think that would really work reasonably well.

Reply:
I would also love to see it enabled by default when CC is deployed, but we should set some throttling to avoid having a significant impact on client applications.

The problem is that there is no way to set the rebalance throttling as a percentage of the current cluster traffic (as seen by the CC cluster model), but only an absolute value in bytes/s.

If we find a way to determine a good throttling value (e.g. CC improvement, Kafka metrics), then I think we could enable this feature by default for both scale up and down. That said, I think it deserves it's own space/proposal.

Reply:
> The problem is that there is no way to set the rebalance throttling as a percentage of the current cluster traffic (as seen by the CC cluster model), but only an absolute value in bytes/s.

How is that configured today for regular rebalances? Isn't there some option in `.spec.cruiseControl` for it?

Reply:
>How is that configured today for regular rebalances? Isn't there some option in .spec.cruiseControl for it?

yes, we have the `.spec.cruiseControl.replicationThrottle` option, which maps to `replication_throttle` of the rebalance endpoint. No limit by default.

https://strimzi.io/docs/operators/latest/configuring#type-KafkaRebalanceSpec-reference

Reply:
Right. But even if the auto-balancing is enabled by default ... the user will still need to enbale CC. So they can configure this and we can respect it, or?

Reply:
The idea is that auto-rebalancing would be enabled by default when CC is enabled. Throttling configuration is not mandatory today. Without throttling, we can have a significant and unexpected impact on cluster performance when scaling the cluster up or down.

IMO, if we want to enable it by default, we either make throttling mandatory, or we figure out a way to set it automatically. The downside of making it mandatory is that the user is in charge of keeping it up to date with the cluster traffic/network capacity, which is not great.

Reply:
I disagree with that part. It is not required today and regular rebalances seem to work fine without it. I don't think the auto-rebalancing will really work differently here.

Reply:
> rebalances seem to work fine without it.

I don't think it works fine for every use case. If you are providing a 24x7 Kafka service with SLAs, you also want to use replicationThrottle. Of course, it would be mandatory only if CC and auto-rebalance is enabled, not always.

Reply:
If you are providing a 24x7 Kafka service with SLAs, you would likely want the throttle for regular rebalances as well, or? That isthe point I'm trying to make - that I do not see much difference in here. Its not like these rebalances would be happening without any knowledge for everyone. The user still needs to trigger the scaling and they would still do the rebalance manually afterward with or without the throttle.

Reply:
> The user still needs to trigger the scaling and they would still do the rebalance manually afterward with or without the throttle.

That's the point, it would be without throttling by default, which is not safe. This is a mistake that the user can do, but the operator shouldn't.

Anyway, this clearly deserve more discussion and thinking, so I would set it as a non goal for this proposal.

Reply:
> I also think that it would be simpler to make it enabled by default in the future (if anyone from the community raises the point) but it's more complicated to go a step back by disabling it by default instead (if the community doesn't like the enabled by default).

You would need to deprecate the current API and add a brand new API for that as the current API has no way to disable it. So no, I think it won't be easier. And that is why I raised it because if we want that, we would need to design the API for that.

Reply:
Agree, so maybe we could adjust the current API to make it simpler in the future? Something like this?

```yaml
autoRebalance:
      enabled: # true or false, it's false by default when not specified
      modes:
          - mode: add-brokers
            settings: my-add-remove-brokers-rebalancing-settings
          - mode: remove-brokers
            settings: my-add-remove-brokers-rebalancing-settings
```

This way the API would just work, we should just "switch" the default behavior for them.

Reply:
No. I guess as a minimum you would need to merge the enable flag into the modes. Otherwise you would move from being able to enable only some modes to enabling all of them or none of them. But it would still be a bit weird pattern ... imagine YAML like this:

```yaml
autoRebalance:
  modes:
    - mode: add-brokers
      settings: my-add-remove-brokers-rebalancing-settings
```

It would be IMHO quite unintuitive that this has all modes enabled. It would also define the state (enabled / disabled by default) for any new mode you would add to the list. So if you would think that if it should be enabled by default one day, we would IMHO need to rethink this completely. You would design the API for it and use for example feature gate if we want to first have it disabled for testing and to get early feedback. If we would expect this to be disabled by default, I think this API would work very nicely.

One thing to consider is that one day we would probably want at least one additional mode for automatic rebalancing outside of scaling -> so that probably should be included in the considerations.

TBH, I'm skeptical we would get enough feedback on this to change our mind. This is something we should decide here. As I said, I would vote of this to be opt-out. But if you want to go with opt-in, I'm ready to respect it.

Reply:
I agree with the issue we have if we want to switch to "enabled by default" at some point, but it would be the same the other way around as well (switching from "enabled to disabled by default").
I don't see any good API that we can have now to allow switching in one direction or the other.

But referring to your example above, if we go with the "enabled by default", for me it's absolutely unintuitive having the autorebalancing running but no `autoRebalance` section in the `Kafka` CR. Also which configuration it would use if you are not specify any reference to a settings/template custom resource? Going with the default options?

Tbh I do still think that it's easier and more intuitive the "disabled by default", where you start with a `Kafka` CR where you don't see `autoRebalance` and it's clear that the autorebalancing is not enabled and then you add that part by configuring modes with corresponding settings to enable it.

I tried to summarize the show scenario and the "disabled by default" looks me more straightforward and understandable from a user perspective even in terms of UX.

| cruiseControl | autoRebalance | disabled by default | enabled by default |
|---|---|---|---|
| No | No | No autorebalancing | No autorebalancing |
| No | Yes | Error! CC not enabled | Error! CC not enabled |
| Yes | No | Autorebalancing for all modes (with default CC options?) | No Autorebalancing |
| Yes | Yes | Enable autorebalancing modes selectively with custom settings </br> 1. modes empty = No autorebalancing or autorebalancing on ALL modes (with default CC options?) </br> 2. modes Not empty = autorebalancing for defined modes | Enable autorebalancing modes selectively with custom settings </br> 1. modes empty = No autorebalancing </br> 2. modes Not empty = autorebalancing for defined modes |

Reply:
Good points for and against having auto-rebalancing enabled/disabled by default. Two thoughts came to mind when reading through this thread:

(1) The `autoRebalance` config can only be manually set when `cruiseControl` is enabled because `autoRebalance` is contained within the `cruiseControl` spec section. Therefore, we can remove the second row from the table above this comment. 

(2) It's worth noting that Cruise Control's self-healing feature which is related to auto-balancing in general is disabled be default. It also relies on the default Cruise Control configs for proposal generation. Following this pattern I am leaning towards requiring users to opt-in to auto-balancing on scaling.

Reply:
Thanks @kyguy ...

(1) You are absolutely right! :-) Anyway I would remove the first one as well, because not having `cruiseControl` field implies no `autoRebalance` field as well. I am going to update the table in the comment (striking the rows).

(2) Referring to the CC self-healing, it is a really good point. I would say our auto-rebalancing goes somehow in that direction (if we can think of a "third" mode running auto-rebalancing during normal cluster operations, not just on scaling). I am towards the "disabled by default" more than before now. Furthermore, the self-healing uses default CC configs which for us could mean avoiding the need for an external settings resource. Anyway, I am still convinced it's useful to allow user to customize config so I would stick with having a settings resource (maybe not a dedicated new one but a `KafkaRebalance` template. I will change the proposal soon for that). Of course, when not specified we can assume the CC default configs will be used.